### PR TITLE
Add Amazon S3 Transfer Acceleration support

### LINKED ISSUE #4252: Add Amazon S3 Transfer Acceleration support
# This is a Feature Proposal

## Description

The framework could optionally enable S3 Transfer Acceleration and use the s3-accelerate endpoint to upload heavy deployment packages.

* **Use case**: heavy deployment packages or "remote" regions

## Additional Data

Feature announcement/description [here](https://aws.amazon.com/blogs/aws/aws-storage-update-amazon-s3-transfer-acceleration-larger-snowballs-in-more-regions/).

I have tested the transfer speed with the AWS CLI, uploading a 43MB file to us-west-2 from western Europe.

### Without transfer acceleration
```bash
$time aws s3 cp ./my-big-file.zip s3://my-bucket/my-big-file.zip
real	0m33.990s
user	0m2.044s
sys	0m0.608s
(peak: 1.7 MiB/s)
```

### With transfer acceleration
```bash
$time aws s3 cp ./my-big-file.zip s3://my-bucket/my-big-file.zip --endpoint-url http://s3-accelerate.amazonaws.com
real	0m6.867s
user	0m1.303s
sys	0m0.346s
(peak: 10.5 MiB/s)
```

## What's needed?

We'd need to enable transfer acceleration (no CloudFormation support yet).
This is how you do it with the CLI:

```bash
aws s3api put-bucket-accelerate-configuration --bucket bucketname --accelerate-configuration Status=Enabled
```

And then simply append the `--endpoint` option when uploading.

Eventually, it could be a simple plugin, but it sounds simple and useful enough to be an option in the core.

### PR DESCRIPTION
**Pull request by alexcasalboni**
## What did you implement:

I have added support for S3 Transfer Acceleration, which requires one API call for enabling it and one S3-client parameter on request.

Closes #4252

As discussed in the issue, we'd like to enable this functionality only when a command line option such as `--aws-s3-accelerate` is specified. I have partially implemented this (see my TODOs), but I'd need some help to properly read options and eventually yaml config too.

## How did you implement it:

I have added a new step called `enableBucketAcceleration` to the `uploadArtifacts` pipeline.

One transfer acceleration is enabled, we also need to enable a `useAccelerateEndpoint` parameter on the S3 Client for each request. Apparently because of an SDK issue (discussed [here](https://github.com/aws/aws-sdk-js/issues/281#issuecomment-314045508)), I also had to force `signatureVersion = 'v2'`.

## How can we verify it:

This should work with any serverless project when you use the  `--aws-s3-accelerate` command line option. 

To verify that it actually works, I recommend deploying a package of at least 5MB to a remote location (e.g. `ap-southeast-2`). The time difference should be noticeable (e.g. from 20s down to 5s).

For example:

```yaml
service: s3-acceleration-test
package:
  exclude:
    - "**/*"
  include:
    - handler.js
    - huge-file.jpg  # make sure this exists (mine was 22MB)
provider:
  name: aws
  runtime: nodejs6.10
  region: ap-southeast-2  # a randomly selected remote region

functions:
  hello:
    handler: handler.hello  # this is the default Node.js hello world function
```

My results (after deploying the first time to create the deployment stack):

```bash
$ time sls deploy
real	1m38.352s
user	0m13.690s
sys	0m2.761s
```

```bash
$ time sls deploy --aws-s3-accelerate
real	0m57.608s
user	0m13.156s
sys	0m2.556s
```

(which is 40s faster!!)

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by alexcasalboni** [comment]
@pmuens @horike37 I have refactored the code, tested it properly, and provided some verification commands. Let me know what you think :)

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 1]
```diff
@@ -572,74 +609,128 @@ describe('AwsProvider', () => {
...
-        const newOptions = {
-          stage: 'optionsStage',
-        };
-        const config = {
-          stage: 'configStage',
-        };
-        serverless = new Serverless(config);
-        serverless.service.provider.stage = 'providerStage';
-        awsProvider = new AwsProvider(serverless, newOptions);
+  describe('#getStage()', () => {
```
Comment (alexcasalboni) [inline lib/plugins/aws/provider/awsProvider.test.js]: @pmuens I've just fixed a minor indentation problem here

**Comment by alexcasalboni** [comment]
@pmuens Travis CI keeps failing because of `nvm`, but everything should be ok now :)

**Comment by pmuens** [comment]
Alright. We'll look into this! Thanks for taking the time to fix this @alexcasalboni 👍

**Comment by alexcasalboni** [comment]
Thank you, @RafalWilinski :)

Something we may improve later on:

1. I am enabling S3 acceleration at each `sls deploy` (i.e. only one POST API call, instead of one GET to check its current config, then an optional POST to enable acceleration if needed) -> I'd guess the GET call is faster and we may optimize this for every deploy but the first one
2. if you deploy WITHOUT S3 acceleration after enabling it, I'm not disabling the accelerated endpoint (i.e. if you don't use it, you don't pay anyway) -> I really don't want to check the S3 config at each deploy when `--aws-s3-accelerate` is not provided, so I would document that you can disable it manually or even leave it enabled without additional costs

Also, I haven't documented the `--aws-s3-accelerate` CLI parameter yet. Can anyone help?

**Comment by RafalWilinski** [comment]
@horike37 Can you double check this? It's LGTM from my side but I'd love to have your input here.

**Comment by horike37** [comment]
@RafalWilinski 
OK. Just a while.

**Comment by alexcasalboni** [comment]
Oh well, I didn't notice that AWS just announced CloudFormation support for S3 Transfer Acceleration :)

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/ReleaseHistory.html

**Comment by horike37** [comment]
Hey @alexcasalboni ! Sorry for delay response:sweat_drops: 

I just took a look into doubts bought up by me and resolved it.
Nevermind the autocomplete issue since that will be added automatically :sweat_smile: 
About multi-provider, could you please add the following to around [here](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/index.js#L47-L48), then it will be shown with only AWS provider.

```
deploy: {
  options: {
    'aws-s3-accelerate': {
      usage: 'Use the new accelerated endpoint for faster deployment',
    },
  },
},
```
Then, about CloudFormation, imo I prefer to use Cloudformation since keeping the consistency of deployment, otherwise, we might open up a can of worms...

## My suggestion is here
### when using this option
- Do nothing if a user specifies `deploymentBucket` in serverless.yml since Serverless should not change the bucket in this case and a warning should be shown.
- Add the property to CF template for creating bucket and update the stack if a deployed service already exists and disabled the option

### when NOT using this option
- Remove the property from CF template for creating bucket and update the stack if a deployed service already exists and  enabled the option

What do you think? 
cc: @serverless/framework-maintainers

**Comment by HyperBrain** [comment]
@horike37 Agree with your proposal. Sound consistent.

**Comment by alexcasalboni** [comment]
Thank you, @horike37. It makes perfect sense to me :)

I am going to:

1. Refactor my code to use CF (add/remove property to template)
2. Add a warning and do nothing if `deploymentBucket` is provided explicitly
3. Add the `aws-s3-accelerate` option to AWS' deploy cmd

**Comment by horike37** [comment]
Thank you for moving ahead @alexcasalboni :100: 
I'm looking forward to completing this functionality :smile: 

Let us know if you need any helps anytime :+1:

**Comment by alexcasalboni** [comment]
@horike37  @pmuens @HyperBrain  I am realizing that the CloudFormation approach works great for new services, but it's tricky for already existing services due to how the framework handles stack updates.

Example 1 (OK):

1. First invocation of `sls deploy --aws-s3-accelerate`
2. `create-stack.json` is correctly configured
3. stack is created with enabled transfer acceleration
4. everything is uploaded using the accelerated endpoint

Example 2 (OK):

1. Service is already deployed with transfer acceleration
2. `sls deploy` is invoked without `--aws-s3-accelerate`
2. `update-stack.json` is corrently updated with the new `AccelerateConfiguration` (disabled)
3. `sls` will update the stack
4. successful deploy (S3 Transfer Acceleration is now disabled)

Example 3 (KO):
1. Service is already deployed WITHOUT transfer acceleration
2. `update-stack.json` is corrently updated with the new `AccelerateConfiguration` (enabled)
3. `sls` will try to upload using the accelerated endpoint
4. BOOM (acceleration is not enabled -> stack hasn't been updated yet)

How would you solve example 3? Enabling Transfer Acceleration via CloudFormation would require one deploy (without using the accelerated endpoint). Only after that first deployment, we can use the accelerated endpoint.

We didn't have this problem by using the S3 API because I was enabling/disabling it at every deployment.

What do you guys think?

**Comment by HyperBrain** [comment]
@alexcasalboni Good point. It could be a solution, that SLS remembers if it enabled the acceleration in the CF template in the _current_ deployment and turns it off for that run. This would catch the KO case and keep the others functional.
Additionally it will handle the 4th case (i.e. when you turn it off again).

**Comment by alexcasalboni** [comment]
@HyperBrain that makes sense. Turning it off should never be a problem.

I'll work on it to make sure it doesn't explode.

But still, most people will be trying this new feature on their existing service and nothing will change until the second deployment (that's not what I'd expect either).

What if we automagically performed two stack updates in that case? The first one would simply update the bucket configuration, and the second one would check/update everything else. Not 100% sure of how to implement it, though.

**Comment by alexcasalboni** [comment]
Update: I've implemented what @horike37 described, following @HyperBrain's suggestion.

Recap:

- CF-only solution
- disabling acceleration always works
- enabling acceleration always works for brand-new services
- enabling acceleration on existing services will require two deployments (a warning is shown during the first one)
- a warning is shown if custom `deploymentBucket` is provided

**Comment by horike37** [comment]
Thanks @alexcasalboni :+1:  We will review and test that again :smile: 

As far as I read the above comment, the implementation looks good :100: 

>What if we automagically performed two stack updates in that case? 
I think no need, it would be sufficient that only a warning is shown during the first deployment.

**Comment by alexcasalboni** [comment]
Thanks, @horike37.

Please note that:

1. I changed the API call used to verify the CloudFormation Stack existence from `describeStackResources` to `describeStacks`, so that I can retrieve the stack Outputs too. Plus, the result of this API call should be smaller in the general case (in my very simple case, it's 43% smaller, and it should be even smaller for services with additional resources)
2. I used a new custom CloudFormation Output named `ServerlessDeploymentBucketAccelerated` to verify whether the currently deployed bucket is already accelerated or not (without any additional API call).

**Comment by horike37** [comment]
Hey @alexcasalboni, thank you for your MUCH efforts and taking a lot of time :+1: :100: 
I just did heavyyyyy testing locally with various situation and great to see working fine perfectly!!
That's awesome :tada: You are our hero :sunglasses: 

This functionality just shipped in the v1.25.0 release!

**Comment by horike37** [comment]
Hey @alexcasalboni, thank you for your MUCH efforts and taking a lot of time :+1: :100: 
I just did heavyyyyy testing locally with various situation and great to see working fine perfectly!!
That's awesome :tada: You are our hero :sunglasses: 

This functionality just shipped in the v1.25.0 release!

**Comment by alexcasalboni** [comment]
@horike37  That's awesome! :)

I'm looking forward to receiving feedback from the community. Keep me posted if any problem shows up!

**Comment by waynerobinson** [comment]
Thanks for adding this feature.

We're using a manually defined S3 bucket (with transfer acceleration enabled), but during a deploy we still get the following message: `Serverless: Not using S3 Transfer Acceleration (1st deploy)`

Is this feature only supported for buckets created by Serverless?

**Comment by horike37** [comment]
>Is this feature only supported for buckets created by Serverless?

Yes, Serverless do nothing for a deployment bucket if you specify `deploymentBucket` in serverless.yml

**Comment by alexcasalboni** [comment]
@waynerobinson you are right, the warning message is a bit misleading. 

@horike37 as long as the transfer acceleration is enabled on the custom bucket, we could still enable it, either with one additional argument or by checking via API.

**Comment by alexcasalboni** [comment]
@waynerobinson I've performed some tests and I found a bug :)

@horike37 let me clarify:

- as long as transfer acceleration has been manually enabled on the custom bucket, `sls deploy --aws-s3-accelerate` will use the accelerated endpoint (you'll get an error if acceleration is not enabled, but everything works if you deploy a new service)
- you should see the warning `S3 Transfer Acceleration will not be enabled on deploymentBucket.` when a custom bucket is used since we won't perform any action on the custom bucket
- The message `Not using S3 Transfer Acceleration (1st deploy)` is supposed to be shown only during the first deploy when a regular deployment bucket is used, but due to a bug it's also shown for custom buckets that were already deployed in the past

I can quickly work on a fix for the next minor release.

**Comment by alexcasalboni** [comment]
@waynerobinson @horike37 I've implemented the fix in #4606 :)

**Comment by horike37** [comment]
@alexcasalboni 
Sounds reasonable https://github.com/serverless/serverless/pull/4293#issuecomment-353748694 :+1:

**Pull request merged**